aptos-move/move-examples/marketplace/sources/token_offer.move

```diff
@@ -0,0 +1,737 @@
+            signer::address_of(seller),
+            token_offer_addr,
+            royalty_payee,
+            royalty_denominator,
+            royalty_numerator,
+        );
+    }
+
+    /// From the coin offer remove appropriate payment for the token and distribute to the seller,
+    /// the creator for royalties, and the marketplace for commission. If there are no more slots,
+    /// cleanup the offer.
+    inline fun settle_payments<CoinType>(
+        seller: address,
+        token_offer_addr: address,
+        royalty_payee: address,
+        royalty_denominator: u64,
+        royalty_numerator: u64,
+    ) acquires CoinOffer, TokenOffer, TokenOfferTokenV1, TokenOfferTokenV2 {
+        assert!(exists<TokenOffer>(token_offer_addr), ENO_TOKEN_OFFER);
+        let token_offer_obj = borrow_global_mut<TokenOffer>(token_offer_addr);
+        assert!(
+            timestamp::now_seconds() < token_offer_obj.expiration_time,
+            error::invalid_state(EEXPIRED),
+        );
+        let price = token_offer_obj.item_price;
+
+        assert!(
+            exists<CoinOffer<CoinType>>(token_offer_addr),
+            error::not_found(ENO_COIN_OFFER),
+        );
+        let coin_offer = borrow_global_mut<CoinOffer<CoinType>>(token_offer_addr);
+        let coins = coin::extract(&mut coin_offer.coins, price);
+
+        let royalty_charge = price * royalty_numerator / royalty_denominator;
+        let royalties = coin::extract(&mut coins, royalty_charge);
+        coin::deposit(royalty_payee, royalties);
+
+        let fee_schedule = token_offer_obj.fee_schedule;
+        let commission_charge = fee_schedule::commission(fee_schedule, price);
+        let commission = coin::extract(&mut coins, commission_charge);
+        coin::deposit(fee_schedule::fee_address(fee_schedule), commission);
+
+        coin::deposit(seller, coins);
+
+        let event = TokenOfferEvent {
+            seller,
+            price,
+            royalties: royalty_charge,
+            commission: commission_charge,
+        };
+        event::emit_event(&mut token_offer_obj.events, event);
+
+        cleanup<CoinType>(object::address_to_object(token_offer_addr));
+    }
+
+    /// Cleanup the offer by deleting it and returning the remaining funds to the token offer
+    /// creator.
+    inline fun cleanup<CoinType>(
+        token_offer: Object<TokenOffer>,
+    ) acquires CoinOffer, TokenOffer, TokenOfferTokenV1, TokenOfferTokenV2 {
+        let token_offer_addr = object::object_address(&token_offer);
+        let CoinOffer<CoinType> { coins } = move_from<CoinOffer<CoinType>>(token_offer_addr);
+        coin::deposit(object::owner(token_offer), coins);
+
+        let TokenOffer {
+            fee_schedule: _,
+            item_price: _,
+            expiration_time: _,
+            delete_ref,
+            events,
+        } = move_from<TokenOffer>(token_offer_addr);
+        event::destroy_handle(events);
+        object::delete(delete_ref);
```

Comment:
are we going to emit another event to track what happened? e.g. bought vs canceled